### PR TITLE
Added indexing for bool tensors and bool Indices

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#18583 Added indexing for bool tensors and bool Indices**
* #18505 Added numpy conversion
* #18166 Bool Tensor for CUDA

-----------
This PR enables bool tensor indexing and indexing with bool indices. This is a part of Bool Tensor feature implementation work. The whole plan looks like this:
1. Storage Implementation [Done]
2. Tensor Creation.
    a) CPU [Done]
    b) CUDA [In review]
3. Tensor Conversions. [In review]
4. Tensor Indexing. [This PR]
5. Tensor Operations.
6. Back compatibility related changes.

TODO: 
as a follow up, we should move nonzero method from TH to Aten to make code cleaner. 

Change:
```
v = torch.tensor([True, False, True], dtype=torch.bool)
boolIndices = torch.tensor([True, False, False], dtype=torch.bool)
v[boolIndices]
-> tensor([True], dtype=torch.bool)


v = torch.randn(5, 7, 3)
boolIndices = torch.tensor([True, False, True, True, False], dtype=torch.bool)
v[boolIndices]
-> 
tensor([[[ 0.5885, -0.3322,  0.7388],
         [ 1.1182,  0.7808, -1.1492],
         [-0.7952,  0.5255, -0.0251],
         [ 0.7128,  0.8099,  1.2689],
         [-0.7018, -1.4733, -0.3732],
         [ 0.4503,  0.4986, -1.1605],
         [ 0.3348, -1.3767, -0.2976]],

        [[-2.0303, -0.4720, -0.1448],
         [-0.1914, -0.6821,  2.0061],
         [-1.0420, -0.1872, -0.3438],
         [ 1.7587, -0.4183, -0.7577],
         [ 1.0094, -0.1950, -0.2430],
         [ 0.1174,  0.3308, -0.5700],
         [ 0.1110, -0.2714,  1.3006]],

        [[-0.1946, -1.4747, -0.4650],
         [-1.0567,  1.0110, -0.2809],
         [ 0.3729, -0.5699,  0.0815],
         [-0.7733, -0.8316,  0.1674],
         [ 1.2000, -0.3745, -1.1679],
         [ 1.7105,  0.9851, -0.1907],
         [-1.1077,  0.2086, -0.0548]]])
```

Differential Revision: [D14673403](https://our.internmc.facebook.com/intern/diff/D14673403)

### CODE REVIEW

File: aten/src/ATen/native/Indexing.cpp [hunk 1]
```diff
@@ -79,19 +79,19 @@ static void checkIndexTensorTypes(TensorList indices) {
...
+          AT_INDEX_ERROR("tensors used as indices must be long, byte or bool tensors");
       }
     }
   }
 }
 
-static std::vector<Tensor> expandByteTensors(const Tensor & self, TensorList indices) {
-  // Expands byte tensors (masks) into the equivalent indexing by LongTensors
+static std::vector<Tensor> expandTensors(const Tensor & self, TensorList indices) {
+  // Expands ByteTensor (masks) or BoolTensor into the equivalent indexing by LongTensors
```
Comment (gchanan) [inline aten/src/ATen/native/Indexing.cpp]: nit: masks should be after both.

File: aten/src/ATen/native/cpu/IndexKernel.cpp [hunk 2]
```diff
@@ -92,7 +92,7 @@ void cpu_index_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef
 }
 
 void index_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef index_stride) {
-  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, iter.dtype(), "index_cpu", [&] {
+  AT_DISPATCH_ALL_TYPES_AND2(at::ScalarType::Half, at::ScalarType::Bool, iter.dtype(), "index_cpu", [&] {
```
Comment (gchanan) [inline aten/src/ATen/native/cpu/IndexKernel.cpp]: this is for indexing where the source is a Bool, not the index, right?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -3673,9 +3673,13 @@
   variants: method, function
 
 - func: nonzero(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
+  cpu_bool: True
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: does this have any effect?  I thought this would only matter if these defined backend-specific dispatch, but neither this nor the one below it does.  Can you check?

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 4]
```diff
@@ -4,6 +4,71 @@
...
+#define IS_NONZERO(val) ((val)!=0)
+#endif
+
+  /* First Pass to determine size of subscripts */
+  TH_TENSOR_APPLY(scalar_t, tensor,
+                  if IS_NONZERO(*tensor_data) {
+                    ++numel;
+                  });
+#ifdef DEBUG
+  THAssert(numel <= LONG_MAX);
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp]: did you change this at all?

File: test/test_indexing.py [hunk 5]
```diff
@@ -36,6 +36,32 @@ def test_step_assignment(self):
...
 
+    def test_bool_indices(self):
+        v = torch.randn(5, 7, 3)
+        boolIndices = torch.tensor([True, False, True, True, False], dtype=torch.bool)
+        self.assertEqual(v[boolIndices].shape, (3, 7, 3))
+        self.assertEqual(v[boolIndices], torch.stack([v[0], v[2], v[3]]))
+
+        v = torch.tensor([True, False, True], dtype=torch.bool)
+        boolIndices = torch.tensor([True, False, False], dtype=torch.bool)
+        intIndices = torch.tensor([1, 0, 0], dtype=torch.uint8)
```
Comment (gchanan) [inline test/test_indexing.py]: maybe call this uint8 indices.

File: aten/src/ATen/native/cpu/IndexKernel.cpp (hunk 2 above)
Comment (izdeby) [reply aten/src/ATen/native/cpu/IndexKernel.cpp]: yes, its for the case when we try to index over a bool tensor.

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp (hunk 4 above)
Comment (izdeby) [reply aten/src/TH/generic/THTensorEvenMoreMath.cpp]: no, i didnt. i just moved it outside of 
#if !defined(TH_REAL_IS_BOOL) /* non bool only part */ as its needed for the indexing.

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (izdeby) [reply aten/src/ATen/native/native_functions.yaml]: Yeah, you are right, this is not needed. Removed

**Comment by ezyang** [comment]
```
v = torch.randn(5, 7, 3)
boolIndices = torch.tensor([True, False, True, True, False], dtype=torch.bool)
v[boolIndices]
```

This example is not very enlightening because I can't see which rows were selected, since they were all uniformly sampled ;)

File: test/test_indexing.py [hunk 6]
```diff
@@ -36,6 +36,32 @@ def test_step_assignment(self):
...
+        y.index_put_((mask, ), y[mask], accumulate=True)
+        self.assertEqual(y, torch.ones(size=(10, 10)))
+
+    def test_multiple_bool_indices(self):
+        v = torch.randn(5, 7, 3)
+        # note: these broadcast together and are transposed to the first dim
+        mask1 = torch.tensor([1, 0, 1, 1, 0], dtype=torch.bool)
+        mask2 = torch.tensor([1, 1, 1], dtype=torch.bool)
+        self.assertEqual(v[mask1, :, mask2].shape, (3, 7))
+
```
Comment (ezyang) [inline test/test_indexing.py]: Something I notice in this test file is we have copied some tests from Numpy. Ostensibly, we removed the boolean tests when we did so. Can we bring those back?

File: test/test_indexing.py [hunk 7]
```diff
@@ -36,6 +36,32 @@ def test_step_assignment(self):
...
+        boolIndices = torch.tensor([True, False, True, True, False], dtype=torch.bool)
+        self.assertEqual(v[boolIndices].shape, (3, 7, 3))
+        self.assertEqual(v[boolIndices], torch.stack([v[0], v[2], v[3]]))
+
+        v = torch.tensor([True, False, True], dtype=torch.bool)
+        boolIndices = torch.tensor([True, False, False], dtype=torch.bool)
+        uint8Indices = torch.tensor([1, 0, 0], dtype=torch.uint8)
+        self.assertEqual(v[boolIndices].shape, v[uint8Indices].shape)
+        self.assertEqual(v[boolIndices], v[uint8Indices])
+        self.assertEqual(v[boolIndices], tensor([True], dtype=torch.bool))
```
Comment (ezyang) [inline test/test_indexing.py]: Can you remind me what the BC-breaking plan for uint8 indices is?

File: test/test_indexing.py [hunk 8]
```diff
@@ -36,6 +36,32 @@ def test_step_assignment(self):
...
+
+    def test_bool_indices_accumulate(self):
+        mask = torch.zeros(size=(10, ), dtype=torch.bool)
+        y = torch.ones(size=(10, 10))
+        y.index_put_((mask, ), y[mask], accumulate=True)
+        self.assertEqual(y, torch.ones(size=(10, 10)))
```
Comment (ezyang) [inline test/test_indexing.py]: Mmm, this doesn't actually test accumulate, right? Because you didn't end up selecting anything in the mask, so `index_put_` is a no-op. (There's nothing wrong with testing this, but it's unexpected from the name of the test method.)

**Review by ezyang — APPROVED** [review]
lgtm

File: test/test_indexing.py (hunk 6 above)
Comment (izdeby) [reply test/test_indexing.py]: @ezyang, this happened before i have joined the team :) should i just look up what tests are there in numpy or is there some straightforward way to see whats missing?

File: test/test_indexing.py (hunk 7 above)
Comment (izdeby) [reply test/test_indexing.py]: There is no plan as of right now. All these changes dont break BC. We will have to come up with a plan on what to do soon

File: test/test_indexing.py (hunk 8 above)
Comment (izdeby) [reply test/test_indexing.py]: I took test_byte_mask_accumulate as an example so thats where the name came from

File: test/test_indexing.py (hunk 6 above)
Comment (gchanan) [reply test/test_indexing.py]: You can look up what's missing from NumPy, probably just based on the name.

**Review by gchanan — APPROVED** [review]
